File: Lab_1/Lab_1.8/Bill.cs
namespace Lab_1._8
{
    public class Bill
    {
        public string LastName { get; private set; }
        public string PhoneNumber { get; private set; }
        public double MinuteRate { get; private set; }
        public double Discount { get; private set; }
        public Time StartTime { get; private set; }
        public Time EndTime { get; private set; }
        public double TotalAmount { get; private set; }
        public class Time
        {
            public uint Hour { get; private set; }
            public uint Minute { get; private set; }
            public uint Second { get; private set; }

            public Time()
            {
            }

            public Time(uint seconds)
            {
                Hour = seconds / 3600;
                Minute = (seconds % 3600) / 60;
                Second = seconds % 60;
            }

            public Time(uint hour, uint minute, uint second)
            {
                Hour = hour;
                Minute = minute;
                Second = second;
            }

            public bool Init(uint hour, uint minute, uint second)
            {
                if (hour < 0 || hour > 23 || minute < 0 || minute > 59 || second < 0 || second > 59)
                {
                    Console.WriteLine("Invalid time format. Please use values within the range: hour[0-23], minute[0-59], second[0-59].");
                    return false;
                }

                Hour = hour;
                Minute = minute;
                Second = second;
                return true;
            }

            public bool Init(string timeString)
            {
                string[] timeParts = timeString.Split(':');
                if (timeParts.Length != 3)
                {
                    Console.WriteLine("Invalid time format. Please use 'hour:minute:second'.");
                    return false;
                }

                if (!uint.TryParse(timeParts[0], out uint hour) || !uint.TryParse(timeParts[1], out uint minute) || !uint.TryParse(timeParts[2], out uint second))
                {
                    Console.WriteLine("Invalid time format. Please use numeric values for hour, minute, and second.");
                    return false;
                }
                Init(hour, minute, second);
                return true;
            }

            public bool Init(uint secondsFromMidnight)
            {
                uint hour = secondsFromMidnight / 3600;
                uint minute = (secondsFromMidnight % 3600) / 60;
                uint second = secondsFromMidnight % 60;

                return Init(hour, minute, second);
            }

            public bool Init(Time otherTime)
            {
                return Init(otherTime.Hour, otherTime.Minute, otherTime.Second);
            }

            public void Read()
            {
                Console.WriteLine("\nSelect the format:");
                Console.WriteLine("1. Initialize with numbers");
                Console.WriteLine("2. Initialize with string");
                Console.WriteLine("3. Initialize with seconds from midnight");
                Console.WriteLine("4. Initialize with default time");
                int Choice = int.Parse(Console.ReadLine());
                switch (Choice)
                {
                    case 1:
                        ReadHour();
                        break;
                    case 2:
                        ReadString();
                        break;
                    case 3:
                        ReadSecondsFromMidnight();
                        break;
                    case 4:
                        ReadCopy();
                        break;
                }
            }

            public void ReadHour()
            {
                uint hours;
                uint minutes;
                uint seconds;
                do
                {
                    Console.WriteLine("Enter hours:");
                    hours = uint.Parse(Console.ReadLine());
                    Console.WriteLine("Enter minutes:");
                    minutes = uint.Parse(Console.ReadLine());
                    Console.WriteLine("Enter seconds:");
                    seconds = uint.Parse(Console.ReadLine());
                } while (!Init(hours, minutes, seconds));
            }

            public void ReadString()
            {
                string input;
                do
                {
                    Console.WriteLine("Enter time in the format 'hour:minute:second':");
                    input = Console.ReadLine();
                } while (!Init(input));

            }

            public void ReadSecondsFromMidnight()
            {
                uint seconds;
                do
                {
                    Console.WriteLine("Enter seconds from midnight:");
                    seconds = uint.Parse(Console.ReadLine());
                } while (!Init(seconds));
            }

            public void ReadCopy()
            {
                Time timeCopy = new Time(0, 0, 0);
                Hour = timeCopy.Hour;
                Minute = timeCopy.Minute;
                Second = timeCopy.Second;
            }

            public void Display()
            {
                Console.Write($"Time:{Hour}:{Minute}:{Second}");
            }

            public uint DifferenceInSeconds(Time otherTime)
            {
                return (otherTime.ToSeconds() - ToSeconds());
            }

            public void AddSeconds(uint seconds)
            {
                uint totalSeconds = ToSeconds() + seconds;
                Hour = totalSeconds / 3600;
                Minute = (totalSeconds % 3600) / 60;
                Second = totalSeconds % 60;

                Hour %= 24;
            }

            public void SubtractSeconds(uint seconds)
            {
                int totalSeconds = (int)ToSeconds() - (int)seconds;
                if (totalSeconds < 0)
                {
                    totalSeconds += 24 * 60 * 60;
                }
                Hour = (uint)(totalSeconds / 3600);
                Minute = (uint)((totalSeconds % 3600) / 60);
                Second = (uint)(totalSeconds % 60);


            }

            public bool IsEqualTo(Time otherTime)
            {
                return ToSeconds() == otherTime.ToSeconds();
            }

            public uint ToSeconds()
            {
                return Hour * 3600 + Minute * 60 + Second;
            }

            public uint ToMinutes()
            {
                return (uint)Math.Round((double)ToSeconds() / 60);
            }

            public override string ToString()
            {
                return $"{Hour}:{Minute}:{Second}";
            }
        }

        public bool Init(string lastName, string phoneNumber, double minuteRate, double discount, Time startTime, Time endTime)
        {
            if (string.IsNullOrEmpty(lastName) || string.IsNullOrEmpty(phoneNumber) || minuteRate <= 0 || discount < 0 || discount > 100 || startTime == null || endTime == null)
            {
                Console.WriteLine("Invalid input! Please check your data.");
                return false;
            }

            LastName = lastName;
            PhoneNumber = phoneNumber;
            MinuteRate = minuteRate;
            Discount = discount;
            StartTime = startTime;
            EndTime = endTime;

            CalculateTotalAmount();
            return true;
        }

        public void Read()
        {
            bool initialized = false;
            do
            {
                Console.WriteLine("Enter details for bill:");
                Console.Write("Last Name: ");
                string lastName = Console.ReadLine();
                Console.Write("Phone Number: ");
                string phoneNumber = Console.ReadLine();
                Console.Write("Minute Rate: ");
                double minuteRate;
                double.TryParse(Console.ReadLine(), out minuteRate);
                Console.Write("Discount: ");
                double discount;
                double.TryParse(Console.ReadLine(), out discount);
                bool ifRepeat = false;
                Time startTime = new Time();
                Time endTime = new Time();
                do
                {
                    Console.WriteLine("Enter start time: ");
                    startTime.Read();
                    Console.WriteLine("Enter end time: ");
                    endTime.Read();
                } while (ifRepeat);
                initialized = Init(lastName, phoneNumber, minuteRate, discount, startTime, endTime);
            } while (!initialized);
        }

        public void Display()
        {
            Console.WriteLine($"Last Name: {LastName}");
            Console.WriteLine($"Phone Number: {PhoneNumber}");
            Console.WriteLine($"Minute Rate: {MinuteRate}");
            Console.WriteLine($"Discount: {Discount}");
            Console.WriteLine("Start Time: " + StartTime.ToString());
            Console.WriteLine("End Time: " + EndTime.ToString());
            Console.WriteLine($"Total Amount: {TotalAmount} UAH");
        }

        private void CalculateTotalAmount()
        {
            double startTimeInMinutes = StartTime.Hour * 60 + StartTime.Minute + StartTime.Second / 60;
            double endTimeInMinutes = EndTime.Hour * 60 + EndTime.Minute + EndTime.Second / 60;

            if (endTimeInMinutes < startTimeInMinutes)
            {
                endTimeInMinutes += 24 * 60;
            }

            double durationInMinutes = endTimeInMinutes - startTimeInMinutes;
            if (durationInMinutes % 1 != 0) { durationInMinutes = (int)durationInMinutes++; }
            double totalCost = durationInMinutes * MinuteRate;
            double discountAmount = totalCost * (Discount / 100);
            TotalAmount = totalCost - discountAmount;
        }

        public override string ToString() => $"Last Name: {LastName}, Phone Number: {PhoneNumber}, Minute Rate: {MinuteRate}, Discount: {Discount}%, StartTime: {StartTime.ToString()}, EndTime: {EndTime.ToString()}, Total Amount: {TotalAmount} UAH";


    }
}

File: Lab_2/Lab_2.8_c++/Bill.cpp
#include "Bill.h"
#include <iostream>
#include <cmath>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <iomanip>
using namespace std;
int Bill::counter = 0;

Bill::Bill() : minuteRate(0), discount(0), totalAmount(0) { counter++; }
Bill::Bill(const std::string& lastName, const std::string& phoneNumber, double minuteRate, double discount, const Time& startTime, const Time& endTime) {
	this->lastName = lastName;
	this->phoneNumber = phoneNumber;
	this->minuteRate = minuteRate;
	this->discount = discount;
	this->startTime = startTime;
	this->endTime = endTime;
	CalculateTotalAmount();
	counter++;
}
Bill::Bill(const Bill& other) {
	lastName = other.lastName;
	phoneNumber = other.phoneNumber;
	minuteRate = other.minuteRate;
	discount = other.discount;
	startTime = other.startTime;
	endTime = other.endTime;
	totalAmount = other.totalAmount;
	CalculateTotalAmount();
	counter++;
}
Bill::~Bill() {
	counter--;
}

void Bill::CalculateTotalAmount() {
	double startTimeInMinutes = startTime.GetHour() * 60 + startTime.GetMinute() + startTime.GetSecond() / 60;
	double endTimeInMinutes = endTime.GetHour() * 60 + endTime.GetMinute() + endTime.GetSecond() / 60;

	if (endTimeInMinutes < startTimeInMinutes) {
		endTimeInMinutes += 24 * 60;
	}

	int durationInMinutes = endTimeInMinutes - startTimeInMinutes;
	if (durationInMinutes % 1 != 0) { durationInMinutes = (durationInMinutes++); }
	double totalCost = durationInMinutes * minuteRate;
	double discountAmount = totalCost * (discount / 100);
	totalAmount = totalCost - discountAmount;
}


Bill& Bill::operator =(const Bill& t)
{
	lastName = t.lastName;
	phoneNumber = t.phoneNumber;
	minuteRate = t.minuteRate;
	discount = t.discount;
	startTime = t.startTime;
	endTime = t.endTime;
	totalAmount = t.totalAmount;
	return *this;
}
std::ostream& operator<<(std::ostream& out, const Bill& bill) {
	out << "Last Name: " << bill.lastName << std::endl;
	out << "Phone Number: " << bill.phoneNumber << std::endl;
	out << "Minute Rate: " << bill.minuteRate << std::endl;
	out << "Discount: " << bill.discount << std::endl;
	out << "Start Time: " << bill.startTime << std::endl;
	out << "End Time: " << bill.endTime << std::endl;
	out << "Total Amount: " << bill.totalAmount << " UAH" << std::endl;
	return out;
}

std::istream& operator>>(std::istream& in, Bill& bill) {
	std::cout << "Enter details for bill:" << std::endl;
	std::cout << "Last Name: ";
	in >> bill.lastName;
	std::cout << "Phone Number: ";
	in >> bill.phoneNumber;
	std::cout << "Minute Rate: ";
	in >> bill.minuteRate;
	std::cout << "Discount: ";
	in >> bill.discount;

	std::cout << "Enter start time:" << std::endl;
	in >> bill.startTime;
	std::cout << "Enter end time:" << std::endl;
	in >> bill.endTime;
	return in;
}
Bill::operator string() const
{
	stringstream ss;
	ss << "Last Name: " << lastName << std::endl;
	ss << "Phone Number: " << phoneNumber << std::endl;
	ss << "Minute Rate: " << minuteRate << std::endl;
	ss << "Discount: " << discount << std::endl;
	ss << "Start Time: " << startTime << std::endl;
	ss << "End Time: " << endTime << std::endl;
	ss << "Total Amount: " << totalAmount << " UAH" << std::endl;
	return ss.str();
}

Bill& Bill::operator ++()
{
	++minuteRate;
	return *this;
}
Bill& Bill::operator --()
{
	--minuteRate;
	return *this;
}
Bill Bill::operator ++(int)
{
	Bill tmp = *this;
	++discount;
	return tmp;
}
Bill Bill::operator --(int)
{
	Bill tmp = *this;
	--discount;
	return tmp;
}


///////////////////   TIME   //////////////////////////////////
int Bill::Time::counter = 0;

Bill::Time::Time() : hour(0), minute(0), second(0) { counter++; }
Bill::Time::Time(unsigned int hour, unsigned int minute, unsigned int second)
	: hour(hour), minute(minute), second(second) {
	counter++;
}
Bill::Time::Time(string timeString) {
	size_t pos1 = timeString.find(':');
	size_t pos2 = timeString.find(':', pos1 + 1);

	if (pos1 == string::npos || pos2 == string::npos || pos1 == 0 || pos2 == timeString.length() - 1) {
		cout << "Invalid time format. Please use 'hour:minute:second'." << std::endl;
	}

	string hourStr = timeString.substr(0, pos1);
	string minuteStr = timeString.substr(pos1 + 1, pos2 - pos1 - 1);
	string secondStr = timeString.substr(pos2 + 1);

	try {
		hour = stoi(hourStr);
		minute = stoi(minuteStr);
		second = stoi(secondStr);

	}
	catch (const invalid_argument& e) {
		cout << "Invalid time format. Please use numeric values for hour, minute, and second." << endl;
	}
	counter++;
}
Bill::Time::~Time() { counter--; }
Bill::Time::Time(unsigned int seconds) {
	hour = seconds / 3600;
	minute = (seconds % 3600) / 60;
	second = seconds % 60;
}
Bill::Time::Time(const Time& other) {
	hour = other.hour;
	minute = other.minute;
	second = other.second;
}

Bill::Time& Bill::Time::operator =(const Time& t)
{
	hour = t.hour;
	minute = t.minute;
	second = t.second;
	return *this;
}
ostream& operator<<(ostream& out, const Bill::Time& time) {
	out << "Time: " << setw(2) << setfill('0') << time.hour << ":" << setw(2) << setfill('0') << time.minute << ":" << setw(2) << setfill('0') << time.second;
	return out;
}
istream& operator >>(istream& in, Bill::Time& x)
{
	std::cout << "\nSelect the format:" << std::endl;
	std::cout << "1. Initialize with numbers" << std::endl;
	std::cout << "2. Initialize with string" << std::endl;
	std::cout << "3. Initialize with seconds from midnight" << std::endl;
	std::cout << "4. Initialize with default time" << std::endl;

	int choice;
	std::cin >> choice;
	string timeString, hourStr, minuteStr, secondStr;
	size_t pos1, pos2;
	switch (choice) {
	case 1:
		std::cout << "Enter hours:";
		in >> x.hour;
		std::cout << "Enter minutes:";
		in >> x.minute;
		std::cout << "Enter seconds:";
		in >> x.second;
		break;
	case 2:
		std::cout << "Enter time in the format 'hour:minute:second':" << std::endl;
		std::cin >> timeString;
		pos1 = timeString.find(':');
		pos2 = timeString.find(':', pos1 + 1);

		if (pos1 == string::npos || pos2 == string::npos || pos1 == 0 || pos2 == timeString.length() - 1) {
			std::cout << "Invalid time format. Please use 'hour:minute:second'." << std::endl;
		}

		hourStr = timeString.substr(0, pos1);
		minuteStr = timeString.substr(pos1 + 1, pos2 - pos1 - 1);
		secondStr = timeString.substr(pos2 + 1);

		try {
			x.hour = stoi(hourStr);
			x.minute = stoi(minuteStr);
			x.second = stoi(secondStr);

		}
		catch (const invalid_argument& e) {
			std::cout << "Invalid time format. Please use numeric values for hour, minute, and second." << endl;
		}
		break;
	case 3:
		unsigned int seconds;
		std::cout << "Enter seconds from midnight:" << endl;
		std::cin >> seconds;
		x.hour = seconds / 3600;
		x.minute = (seconds % 3600) / 60;
		x.second = seconds % 60;
		break;
	case 4:
		x.hour = 0;
		x.minute = 0;
		x.second = 0;
		break;
	default:
		std::cout << "Invalid choice" << std::endl;
		break;
	}
	return in;
}
Bill::Time::operator string() const
{
	stringstream ss;
	ss << setw(2) << setfill('0') << hour << ":" << setw(2) << setfill('0') << minute << ":" << setw(2) << setfill('0') << second;
	return ss.str();
}

Bill::Time& Bill::Time::operator ++()
{
	++hour;
	return *this;
}
Bill::Time& Bill::Time::operator --()
{
	--hour;
	return *this;
}
Bill::Time Bill::Time::operator ++(int)
{
	Bill::Time tmp = *this;
	++minute;
	return tmp;
}
Bill::Time Bill::Time::operator --(int)
{
	Bill::Time tmp = *this;
	--minute;
	return tmp;
}

unsigned int Bill::Time::DifferenceInSeconds(Time otherTime) {
	return otherTime.ToSeconds() - ToSeconds();
}

void Bill::Time::AddSeconds(unsigned int seconds) {
	unsigned int totalSeconds = ToSeconds() + seconds;
	hour = totalSeconds / 3600;
	minute = (totalSeconds % 3600) / 60;
	second = totalSeconds % 60;

	hour %= 24;
}

void Bill::Time::SubtractSeconds(unsigned int seconds) {
	int totalSeconds = static_cast<int>(ToSeconds()) - static_cast<int>(seconds);
	if (totalSeconds < 0) {
		totalSeconds += 24 * 60 * 60;
	}
	hour = static_cast<unsigned int>(totalSeconds / 3600);
	minute = static_cast<unsigned int>((totalSeconds % 3600) / 60);
	second = static_cast<unsigned int>(totalSeconds % 60);
}

bool Bill::Time::IsEqualTo(Time otherTime) {
	return ToSeconds() == otherTime.ToSeconds();
}

unsigned int Bill::Time::ToMinutes() {
	return ToSeconds() / 60;
}

unsigned int Bill::Time::ToSeconds() {
	return (hour * 3600 + minute * 60 + second);
}

